### ISSUE TITLE
[helion] atomic_add gets bug

### ISSUE THREAD

**Issue by Stonepia** [intel/intel-xpu-backend-for-triton#6193]
### Describe the bug

We got error when running the helion ut. The atomic_add seems be additionally triggered, so the results are overcount and stacked.

## Original Bugs in Helion

```bash
# In Helion folder
python -m pytest -k test_error_in_non_taken_branch test/test_control_flow.py
```

<details>
<summary><code>test_error_in_non_taken_branch</code></summary>

```python
def test_error_in_non_taken_branch(self):
        def mul_relu_block_back_spec(x, y, dz):
            z = torch.relu(x * y[:, None])
            grad_x, grad_y = torch.autograd.grad(z, [x, y], dz, retain_graph=True)
            return grad_x, grad_y
    
        @helion.kernel(config=helion.Config(block_sizes=[32, 32]))
        def mul_relu_block_backward_kernel(
            x: torch.Tensor,
            y: torch.Tensor,
            dz: torch.Tensor,
            use_atomics: hl.constexpr = False,
        ):
            # Get tensor sizes
            m, n = x.shape
            # Create output tensor for gradients
            dx = torch.empty_like(x)
    
            if use_atomics:
                dy = torch.zeros_like(y)
            else:
                dy = torch.empty_like(x)
    
            # Use Helion to tile the computation
            for tile_i, tile_j in hl.tile([m, n]):
                # Get input tiles
                x_tile = x[tile_i, tile_j]
                y_tile = y[tile_i]
                dz_tile = dz[tile_i, tile_j]
    
                # For ReLU, gradient is 1 where input > 0, 0 otherwise
                relu_mask = (x_tile * y_tile[:, None]) > 0
                # Chain rule: dx = dz * relu_grad * y
                relu_grad = torch.where(relu_mask, 1, 0)
                dx[tile_i, tile_j] = dz_tile * relu_grad * y_tile[:, None]
    
                # Chain rule: dy = dz * relu_grad * x -> backwards of broadcast(sum)
                if use_atomics:
                    local_dy_grad = torch.sum(dz_tile * relu_grad * x_tile, dim=1)
                    hl.atomic_add(dy, [tile_i], local_dy_grad)
                else:
                    local_dy_grad = dz_tile * relu_grad * x_tile
                    dy[tile_i, tile_j] = local_dy_grad
    
            if use_atomics:
                return dx, dy
            return dx, dy.sum(axis=-1)
    
        x = torch.randn(512, 1024, device=DEVICE, requires_grad=True)
        y = torch.randn(512, device=DEVICE, requires_grad=True)
        dz = torch.randn(512, 1024, device=DEVICE)
        expected = mul_relu_block_back_spec(x, y, dz)
        torch.testing.assert_close(
            mul_relu_block_backward_kernel(x, y, dz, False),
            expected,
            atol=1e-4,
            rtol=1e-4,
        )
        code, output = code_and_output(
            mul_relu_block_backward_kernel,
            (x, y, dz, True),
        )
>       torch.testing.assert_close(
            output,
            expected,
            atol=1e-4,
            rtol=1e-4,
        )
E       Mismatched elements: 512 / 512 (100.0%)
E       Greatest absolute difference: 154.40457153320312 at index (221,) (up to 0.0001 allowed)
E       Greatest relative difference: 2.000037431716919 at index (93,) (up to 0.0001 allowed)
```

</details>

## Standalone reproducer

[reproducer_xpu_atomic_bug.py](https://github.com/user-attachments/files/25534727/reproducer_xpu_atomic_bug.py)

The reproducer results are as below. 

Every kernel uses a flat 1-D grid of nm × nn programs where `nm = M / BLOCK_M`, `nn = N / BLOCK_N`. Each program is responsible for one BLOCK_M × BLOCK_N tile. The expected contract:

- Each program loads its tile of x (shape BLOCK_M × BLOCK_N)
- Reduces along the column axis → a length-BLOCK_M vector val
- Atomically accumulates val into the BLOCK_M output rows it owns

Because there are nn column-tile programs per row tile, after all programs complete `dy[i]` should equal `x[i, :].sum() = N` (**all ones, so the expected result is 128**).

| Condition | Extra firings |
|-----------|---------------|
| A — nm ≥ 2 and nn ≥ 2 (both row and column tile counts > 1) | +1× |
| B — nm ≥ 2 and a 1-D row-indexed tl.load(ptr + row_indices) is in the same kernel | +1× |

```
$ python reproducer_xpu_atomic_bug.py

========================================================================
XPU tl.atomic_add over-count reproducer
========================================================================

--- Case A: atomic_add with 2-D tiled grid, no 1-D load ---
  Expected: dy[i] == N (each row sums N ones over all column tiles)
  BM=128 BN=128                               nm=1 nn=1  expected=128  actual=128  [OK]
  BM=64 BN=128                                nm=2 nn=1  expected=128  actual=128  [OK]
  BM=32 BN=128                                nm=4 nn=1  expected=128  actual=128  [OK]
  BM=32 BN=64                                 nm=4 nn=2  expected=128  actual=128  [OK]
  BM=32 BN=32                                 nm=4 nn=4  expected=128  actual=256  [BUG (over-count 2.00x)]

--- Case B: same kernel + extra 1-D row load ---
  Expected: dy[i] == N (y==1, so val == row_sum(x) == N per tile)
  BM=128 BN=128                               nm=1 nn=1  expected=128  actual=128  [OK]
  BM=64 BN=128                                nm=2 nn=1  expected=128  actual=256  [BUG (over-count 2.00x)]
  BM=32 BN=128                                nm=4 nn=1  expected=128  actual=256  [BUG (over-count 2.00x)]
  BM=32 BN=64                                 nm=4 nn=2  expected=128  actual=256  [BUG (over-count 2.00x)]
  BM=32 BN=32                                 nm=4 nn=4  expected=128  actual=384  [BUG (over-count 3.00x)]

--- Case A: atomic_add with 2-D tiled grid, no 1-D load ---
  Expected: dy[i] == N (each row sums N ones over all column tiles)
BM=128 BN=128  nm=1 nn=1  → OK       # neither condition met
BM=64  BN=128  nm=2 nn=1  → OK       # condition A requires nn≥2 too — not met
BM=32  BN=128  nm=4 nn=1  → OK       # same
BM=32  BN=64   nm=4 nn=2  → OK       # BN=64 means nn=2 but BM≠BN; still OK here
BM=32  BN=32   nm=4 nn=4  → 2×  BUG  # nm≥2 AND nn≥2 → condition A fires

--- Case B: same kernel + extra 1-D row load ---
  Expected: dy[i] == N (y==1, so val == row_sum(x) == N per tile)
BM=128 BN=128  nm=1 nn=1  → OK       # nm=1, neither condition met
BM=64  BN=128  nm=2 nn=1  → 2×  BUG  # nm≥2 + 1-D load → condition B fires alone
BM=32  BN=128  nm=4 nn=1  → 2×  BUG  # same
BM=32  BN=64   nm=4 nn=2  → 2×  BUG  # condition B fires; condition A would need BN=32
BM=32  BN=32   nm=4 nn=4  → 3×  BUG  # condition A (+1) AND condition B (+1) both fire


Bug summary:
  Case A (atomic + 2-D load, no 1-D load):
    BUG triggers when both BLOCK_M=32 and BLOCK_N=32 (nm=4, nn=4)
    -> atomic_add fires 2x per program, result is doubled
  Case B (atomic + 2-D load + 1-D row load):
    BUG triggers whenever nm >= 2 (regardless of nn)
    -> atomic_add fires 2x per program when nm >= 2, nn < 4
    -> atomic_add fires 3x per program when nm >= 4, nn >= 4 (BLOCK_M=BLOCK_N=32)
  nm == 1 (BLOCK_M == M): always correct in both cases
```

### Environment details

Triton: 3.7.0+git0a28f4d7 (Mon Feb 23)
Helion: 8e6de8dec134192a6003d0211ef78c121c6e2c0c ( Mon Feb 23 )

sycl-ls:
[level_zero:gpu][level_zero:0] Intel(R) oneAPI Unified Runtime over Level-Zero, Intel(R) Data Center GPU Max 1550 12.60.7 [1.6.33578+42]

**Comment by etiotto** [comment]
Given:

```
rows = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
cols = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
x = tl.load(x_ptr + rows[:, None] * N + cols[None, :])  # 2D load
val = tl.sum(x, axis=1)                                 # reduce → 1D
tl.atomic_add(dy_ptr + rows, val)                       # 1D atomic
```

The IR after removing layout conversion is (notice the presence of 2 `atomic_rmw` operations): 

```
#blocked = #ttg.blocked<{sizePerThread = [1, 4], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0]}>

tt.func @kernel_atomic_no_1d_load(%arg0: !tt.ptr<f32>, %arg1: !tt.ptr<f32>) {
    ...
    // rows rematerialized with TWO different slice encodings:
    %8 = arith.addi ... : tensor<32xi32, #ttg.slice<{dim = 1, parent = #blocked}>>   // ← for 2D path
    %9 = arith.addi ... : tensor<32xi32, #ttg.slice<{dim = 0, parent = #blocked}>>   // ← cloned for atomic path

    // reduce result
    %22 = "tt.reduce"(...) -> tensor<32xf32, #ttg.slice<{dim = 1, parent = #blocked}>>

    // FIRST atomic: uses slice<dim=1> encoding (from the reduce output)
    %25 = tt.addptr ..., %8 : tensor<32x!tt.ptr<f32>, #ttg.slice<{dim = 1, ...}>>
    %28 = tt.atomic_rmw fadd, relaxed, gpu, %25, %22, %cst_0
        : (tensor<32x!tt.ptr<f32>, #ttg.slice<{dim = 1, parent = #blocked}>>,
           tensor<32xf32,          #ttg.slice<{dim = 1, parent = #blocked}>>,
           tensor<32xi1,           #ttg.slice<{dim = 1, parent = #blocked}>>)

    // SECOND atomic (DUPLICATE): uses slice<dim=0> encoding
    %26 = tt.addptr ..., %9 : tensor<32x!tt.ptr<f32>, #ttg.slice<{dim = 0, ...}>>
    %27 = ttg.convert_layout %22 : ... -> tensor<32xf32, #ttg.slice<{dim = 0, ...}>>
    %29 = tt.atomic_rmw fadd, relaxed, gpu, %26, %27, %cst
        : (tensor<32x!tt.ptr<f32>, #ttg.slice<{dim = 0, parent = #blocked}>>,
           tensor<32xf32,          #ttg.slice<{dim = 0, parent = #blocked}>>,
           tensor<32xi1,           #ttg.slice<{dim = 0, parent = #blocked}>>)

    tt.return
}
```
The `rows` variable feeds two paths:
- 2D load path: rows → convert_layout to slice<dim=1> → expand_dims → 2D broadcast → load → reduce
- Atomic pointer path: rows → addptr → atomic_rmw

Therefore `rows` was rematerialized twice, and each rematerialization produces its own complete chain (including a separate tt.atomic_rmw) so that operations fires twice per program, doubling the result.

**Comment by etiotto** [comment]
The Intel `RemoveLayoutConversions` pass has two phases:

- LayoutPropagation (forward): propagates encodings from anchored ops
- LayoutRematerialization (backward): hoists convert_layout ops by cloning their operand subgraphs with different encodings


The issue here is that during backward rematerialization, the pass tries to eliminate a convert_layout on the 2D load path. It walks backward to `rows` and needs to rematerialize it with a slice<dim=1> encoding. It then calls `propagateToUsers()` to forward-propagate this new encoding to all users of the rematerialized value.

Because `AtomicRMWOp` has the `SameOperandsAndResultEncoding` trait,  `propagateToUsers()` sets the atomic's encoding to the new slice encoding. Then `rewriteSlice()` clones it with that encoding, creating a second `tt.atomic_rmw` in the IR.
